feat : 그룹 랭킹 조회 API에 변동 정보 추가

rladmstn opened the pull request:
## 📌 Related Issue
<!-- issue 링크 -->
https://github.com/GAMZA-BAT/algohub-server/issues/111

## 🚀 Description
<!-- 작업 내용 -->
- 랭킹 조회 API들에 대해 랭킹 변동 사항을 추가했습니다.
- 랭킹의 자세한 구현 내용은 아래와 같습니다!

- 랭킹 관련 정보를 가질 `Ranking` 도메인을 새로 생성했습니다.
```java
private int solvedCount; // 맞은 문제 수
private int rank; // 현재 랭킹
private int rankDiff; // 현재 랭킹까지 오는데 발생한 변동 수준
private double score; // (마감날짜 - 최초의 맞은 풀이 제출 시간) 합산한 점수 
```
- `ranking`은 `GroupMember` 엔티티와 1:1 매핑됩니다.
- score의 경우, 그룹 멤버가 특정 문제에 대해 최초로 맞은 풀이를 제출했을 때 갱신됩니다.
- score의 갱신 값은 `score += (마감 날짜 - 최초의 맞은 풀이 제출 시간)sec * N`이 됩니다.
   - `N` : score 값이 과도하게 커지는 것을 방지하기 위한 임의의 수로, 현재는 `1e-4`로 지정했습니다.
   - 랭킹을 정렬할 때 문제 푼 수가 동일할 경우, score 값을 2차 기준으로 하여 순위를 매깁니다.
   - (원래는 float 타입으로 할까 했는데, score 단위가 sec인지라 최대값이 더 클 double을 사용하기로 결정했습니다.)
  
---

### 1. 최초의 맞은 풀이 생성 시 score 업데이트 및 랭킹 업데이트
- 최초의 맞은 풀이가 생성될 때, score를 계산해 해당 ranking의 score 값을 업데이트 합니다.
- score 업데이트 후, 해당 멤버가 포함된 그룹의 랭킹을 업데이트 합니다.

### 2. 그룹 생성 시 Ranking 생성
- 유저가 그룹을 새로 생성할 때, 그룹장에 대한 새로운 `Ranking`을 생성합니다.

### 3. 그룹 코드를 사용한 그룹 참여 시 Ranking 생성
- 유저가 그룹 코드를 사용해 그룹을 참여할 때, 생성되는 `GroupMember`에 매핑된 `Ranking`을 생성합니다.
- 현재 랭킹으로는 전체 멤버 수, 즉 마지막 등수로 초기화 됩니다.

### 4. 그룹 탈퇴 시 Ranking 제거 및 업데이트
- 그룹에서 탈퇴하는 멤버의 Ranking을 제거합니다.
- `deleteMemberFromStudyGroup` 메서드가 정상적으로 완료되면 랭킹을 업데이트 하도록 합니다.
- 랭킹 업데이트는 공통 관심 사항으로 판단하여 AOP로 실행되도록 구현했습니다.

### 5. 그룹 삭제 시 모든 Ranking 제거
- 그룹장이 해당 그룹을 삭제하면 그룹에 포함되는 모든 Ranking을 제거합니다.

---

- 공통 관심 사항으로 AOP를 확실하게 분리하고자, 같은 도메인의 기능들이지만 `RankingUpdateService`와 `RankingService`로 분리했습니다.
- `Ranking` 도메인이 꽤 기능이 많은 것 같아서 기존의 `studygroup` 디렉토리를 `group` 디렉토리로 만들고, 그 안에서 `ranking`,`studygroup`로 파생되도록 구성했습니다.


## 📢 Review Point
<!-- 코드리뷰 할 때 더 집중해서? 봐주면 좋을 포인트들 -->
- 패키지 구조를 좀 변경한지라 diff로 여기저기가 많이 찍힙니다. 집중적으로 봐주시면 좋을 파일들은 `RankingService`, `RankingUpdateService`, `StudyGroupService`, `Ranking` 입니다.
- 설계도 여러 번 수정되고 여기저기 수정 해야 할 곳들이 많았던지라, 제가 놓친 것들이 꽤 있을 것 같습니다.. 
- 로직적으로 맞지 않은 부분이나 개선할 부분, 엣지 케이스 같은 게 있다면 알려주세요!

## 📚Etc (선택)
<!-- 작업 도중 생긴 특이사항, 또는 고려할 것들(?) -->

rladmstn committed:
feat : 실시간 랭킹 변동 정보 반영하도록 로직 수정

 맞은 풀이가 추가될 때마다 랭킹을 업데이트하는 로직 추가

rladmstn committed:
refactor : 맞은 풀이 확인하는 메서드 isCorrect 추출

rladmstn committed:
feat : 랭킹 정보 반영할 Rank 엔티티 추가

rladmstn committed:
feat : 그룹 생성 시 Rank 생성 로직 추가

rladmstn committed:
feat : Rank 객체 조회에 대한 에러 객체 및 handler 추가

rladmstn committed:
refactor : Rank 엔티티 현재 랭크로 rank -> currentRank 필드명 수정

Rank 필드에 현재 랭킹 정보로 rank -> currentRank로 필드명 수정

rladmstn committed:
refactor : Rank 엔티티의 rankDiff 타입을 int -> String으로 변경

rladmstn committed:
refactor : GroupMember 엔티티에서 rank 관련 필드 및 메서드 제거

rladmstn committed:
feat : Rank QueryDsl 메서드 추가

rladmstn committed:
feat : Rank 엔티티 사용한 실시간 랭킹 변동 및 갱신 로직 구현

rladmstn committed:
test : 랭킹 업데이트 성공 테스트 작성

rladmstn committed:
test : 랭킹 업데이트 실패 테스트 작성

랭킹 업데이트 시 유저 랭킹 정보 조회가 실패한 케이스 테스트 작성

rladmstn committed:
feat : 랭킹 response DTO에 rankDiff 추가

랭킹 변동 지표를 나타내는 rankDiff 추가

rladmstn committed:
feat : 전체 랭킹 조회 및 Top 3 랭킹 조회 시 로직 수정

전체 랭킹 및 Top 3 랭킹 조회 시 List<Rank>를 조회하도록 수정

rladmstn committed:
test : 전체 랭킹 조회 및 Top 3 랭킹 조회 테스트 작성

rladmstn committed:
Resolved conflict

rladmstn committed:
refactor : 랭킹 확인 권한 검증 실패 시 에러 객체 수정

UserValidation -> GroupMemberValidation으로 FORBIDDEN return

rladmstn committed:
test : 랭킹 조회 테스트 시 권한 없는 경우 error 필드명 수정

rladmstn committed:
test : 전체 랭킹 조회 및 Top 3 랭킹 조회 Controller 테스트 작성

rladmstn committed:
feat : 랭킹 업데이트 시, 문제 푼 수가 같으면 먼저 가입한 순서대로 정렬 추가

rladmstn committed:
test : 가입 날짜 따른 랭킹 업데이트 위한 테스트 수정

rladmstn committed:
refactor : 랭킹 엔티티 Rank -> Ranking으로 이름 수정

MySql 예약어 처리로 인한 Rank -> Ranking 이름 수

rladmstn committed:
refactor : QueryDsl QRank -> QRanking으로 이름 수정

rladmstn committed:
feat : 멤버로 그룹 참여 시 Ranking 엔티티 추가

rladmstn committed:
fix : 맞은 풀이가 존재하는지 확인하는 메서드에서 점수 형태는 포함 안되는 버그 해결

기존에는 "맞았습니다!!"로만 맞은 풀이를 확인했는데, 점수형태도 맞은 풀이가 되므로 수정

rladmstn committed:
feat : 그룹 탈퇴 및 삭제 시, Ranking 엔티티 삭제

rladmstn committed:
refactor : group 디렉토리 추가 및 파생 디렉토리로 ranking, studygroup 추가

디렉토리 분리 및 이동에 따른 랭킹 조회 관련 API의 Controller, Service 분리

rladmstn committed:
test : 랭킹 관련 API는 ranking 디렉토리로 이동 및 분리

rladmstn committed:
feat : RankingUpdateService와 RankingService 분리 및 멤버 삭제 시 RankingUpdate AOP 적용

rladmstn committed:
test : RankingUpdateService와 RankingService 분리, 멤버 삭제 시 RankingUpdate AOP 적용에 대한 테스트 작성

rladmstn committed:
feat : Ranking에 score 필드 추가

rladmstn committed:
feat : 처음 맞은 풀이 제출 시, score에 점수 업데이트 로직 추가

rladmstn committed:
test : Ranking update 관련 AOP 동작 테스트와 update 로직 테스트 분리

rladmstn committed:
feat : 랭킹 업데이트 시, score 따른 2차 정렬 로직 추가

rladmstn committed:
test : 불필요한 Problem 객체 제거

rladmstn committed:
feat : 랭킹 업데이트 및 score 업데이트 메서드 SolutionService -> RankingService로 이동

rladmstn committed:
feat : SCORE_SCALING_FACTOR 상수 private에서 public으로 수정

rladmstn committed:
refactor : updateScoreAndRanking 메서드를 updateRanking, updateScore로 분리

rladmstn committed:
test : 랭킹 점수 업데이트 테스트 작성

rladmstn committed:
refactor : 불필요한 DI 제거

rladmstn committed:
test : RankingControllerTest 불필요한 Mocking 제거

rladmstn committed:
test : RankingUpdateServiceAOPTest 불필요한 객체 제거

hwangjokim commented:
이것 조금 꼼꼼한 리뷰 필요하면 저도 볼게요

rladmstn commented:
> 이것 조금 꼼꼼한 리뷰 필요하면 저도 볼게요

안바쁘시다면야 봐주심 좋습니다

hwangjokim commented on src/main/java/com/gamzabat/algohub/exception/CustomExceptionHandler.java:
```diff
@@ -115,4 +116,9 @@ protected ResponseEntity<Object> handler(CheckNicknameValidationException e) {
 	}
+
+	@ExceptionHandler(CannotFoundRankingException.class)
+	protected ResponseEntity<Object> handler(CannotFoundRankingException e) {
```
[minor] Object -> ErrorResponse로 변경해주세요

hwangjokim commented on src/main/java/com/gamzabat/algohub/feature/group/ranking/domain/Ranking.java:
```diff
@@ -0,0 +1,57 @@
+		this.rankDiff = newRankDiff;
+	}
+
+	public void updateScore(double newScore) {
```
네이밍이 업데이트인데, 실제로 해주는건 기존값에 더해주는거여서 조금 헷갈려요. `increase` 같은건 어떨까요?

rladmstn replied:
흠 `increaseScore`가 낫겠네요! 피드백 감삼다

hwangjokim commented on src/main/java/com/gamzabat/algohub/feature/group/ranking/service/RankingService.java:
```diff
@@ -0,0 +1,104 @@
+				r.getMember().getUser().getNickname(),
+				r.getMember().getUser().getProfileImage(),
+				r.getCurrentRank(),
+				(long)r.getSolvedCount(),
```
이것 DTO상에서는 long이여서 캐스팅 된 것 같은데요, 둘이 타입을 맞춰주지 않아야 하는 이유가 있을까요?

rladmstn replied:
그럼 DTO를 실수한 것 같네요.. int가 맞습니다 캐치 감삼다

hwangjokim commented on src/main/java/com/gamzabat/algohub/feature/group/ranking/service/RankingService.java:
```diff
@@ -0,0 +1,104 @@
+
+	private List<GetRankingResponse> getRankingResponse(List<Ranking> ranking) {
+		return ranking.stream().map(r -> new GetRankingResponse(
+				r.getMember().getUser().getNickname(),
```
이쪽 보면 getMember, getUser를 항상 해옵니다. 그런데 repository상으로 봤을때는 fetch join하지 않는 것 같는데, 주요한 이유가 없다면 fetch join하는건 어떨까요

rladmstn replied:
생각 못한 부분이었습니다 fetch join 좋네요!

hwangjokim replied:
요것 아직 반영 안되었네요!
아마 그냥 아직 작업 시작 안하신 것 같은데, 혹시 누락일까봐 리마인드 드립니다

rladmstn replied:
아 어제 뒤늦게 발견하고 작업했는데 커밋을 안했군요..? 지금 푸시 했습니다-

hwangjokim commented on src/main/java/com/gamzabat/algohub/feature/group/ranking/service/RankingService.java:
```diff
@@ -0,0 +1,104 @@
+			.stream()
+			.filter(r -> r.getSolvedCount() != 0)
+			.sorted(Comparator.comparing(Ranking::getCurrentRank))
+			.toList();
```
[nit] `.limit(3)`을 하면 밑에 있는 If를 제거 할 수 있을것같아요

hwangjokim commented on src/main/java/com/gamzabat/algohub/feature/group/ranking/service/RankingService.java:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	private double calculateNewScore(LocalDate problemEndDate, LocalDateTime solvedDateTime) {
+		LocalDateTime endDateTime = problemEndDate.atTime(LocalTime.MAX);
```
신기한 타입 많네요..

hwangjokim commented on src/main/java/com/gamzabat/algohub/feature/group/ranking/service/RankingUpdateService.java:
```diff
@@ -0,0 +1,58 @@
+		});
+		for (Ranking r : rankings) {
+			int originRank = r.getCurrentRank();
+			int newRank = rankings.indexOf(r) + 1;
```
이거 `for (i = 0; i< ranking.size()...`로 변경하고, newRank는 단순히 `i+1`로 가져가면 O(N^2)를 O(N)으로 줄일 수 있을 것 같아요.

hwangjokim reviewed (commented):
좋습니다~ 고생많으셨어요.
@sh0723  @s-hwan 내일 리뷰 시작하면 요거 먼저 봐주세요. diff가 커서 먼저 머지시키는게 좋을것같아요

rladmstn committed:
Resolve conflict

rladmstn committed:
refactor : 자잘한 오류 수정 및 피드백 반영

- CustomExceptionHandler ErrorResponse 타입 지정
- 점수 갱신 updateScore -> increaseScore 네이밍 수정
- Top3 랭킹 조회 시, limit(3) 반영
- newRank 갱신 시, for문 개선

sh0723 commented on src/main/java/com/gamzabat/algohub/feature/group/ranking/service/RankingUpdateService.java:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	@AfterReturning(
+		value = "execution(* com.gamzabat.algohub.feature.group.studygroup.service.StudyGroupService.deleteMemberFromStudyGroup(..)) && args(.., studyGroup)"
```
이부분은 어떤 역할을 하는건지 설명해주실 수 있나요 ?? 지피티한테 물어봐도 잘 모르겠슴다 ㅜ.ㅜ

sh0723 replied:
이거만 혹시 답장 가능할까요 은수쒸?? @rladmstn

sh0723 replied:
일단 머지하겟슴다..!

hwangjokim reviewed (approved):
LGTM

rladmstn committed:
refactor : 그룹 랭킹 조회 시 fetch join 적용

s-hwan reviewed (approved):
고생하셨어용 코드를 가슴으로 공감하고 갑니다 언젠간 머리로 이해할 수 있는 날이 오겠지요..?

rladmstn committed:
Resolve conflict

rladmstn committed:
refactor : NotificationMessage format warning 제거를 위한 타입 수정

- String... args 에서 Object... args로 수정
